### PR TITLE
Airflow db

### PR DESCRIPTION
**Pull request by DomMarq**
Added functionality to do basic SQL requests to the airflow database housed in the istio-release-ui project, both through a standard function and through an HTTP GET request.

### CODE REVIEW

File: rest_api.py [hunk 1]
```diff
@@ -1,11 +1,13 @@
 """REST API."""
-
-import json
-from file_adapter import FileAdapter
-from filter import filter_releases
-from filter import sort
-from flask_restful import reqparse
-from flask_restful import Resource
+import sys
+sys.path.append('../')
```
Comment (sebastienvas) [inline rest_api.py]: why is this required ?

File: rest_api.py [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
+import json   # pylint: disable=g-import-not-at-top, g-bad-import-order
+from file_adapter import FileAdapter   # pylint: disable=g-import-not-at-top
+from filter import filter_releases   # pylint: disable=g-import-not-at-top
```
Comment (sebastienvas) [inline rest_api.py]: filter seems to be a keyword in python, you might want to rename the modules to filters.

File: rest_api.py [hunk 3]
```diff
@@ -1,11 +1,13 @@
...
-
-import json
-from file_adapter import FileAdapter
-from filter import filter_releases
-from filter import sort
-from flask_restful import reqparse
-from flask_restful import Resource
+import sys
+sys.path.append('../')
+import json   # pylint: disable=g-import-not-at-top, g-bad-import-order
```
Comment (sebastienvas) [inline rest_api.py]: can you fix the import order instead of disabling pylint ?

File: rest_api.py [hunk 4]
```diff
@@ -101,10 +103,22 @@ def get(self):
     return json.dumps(response)
 
 
+class AirflowDB(Resource):
```
Comment (sebastienvas) [inline rest_api.py]: you might want to reuse the connection here instead of creating a db connection for each request.

File: airflow_connector.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
+"""Connects to the Cloud SQL database."""
```
Comment (hklai) [inline airflow_connector.py:1]: Add copyright stuff

File: airflow_connector.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    # Connect using the unix socket located at
+    # /cloudsql/cloudsql-connection-name.
+    cloudsql_unix_socket = os.path.join(
+        '/cloudsql', CLOUDSQL_CONNECTION_NAME)
+
+    db = MySQLdb.connect(
+        unix_socket=cloudsql_unix_socket,
+        user=CLOUDSQL_USER,
+        passwd=CLOUDSQL_PASSWORD,
+        db='airflow-db')
```
Comment (hklai) [inline airflow_connector.py]: is the db name always "airflow-db"?

File: rest_api.py (hunk 1 above)
Comment (DomMarq) [reply rest_api.py]: appending the path to sys allows python to import files which are not in the same directory, and allows the tests to be run from the tests/ folder, without the test script having to be moved to the same directory

File: rest_api.py (hunk 4 above)
Comment (DomMarq) [reply rest_api.py]: I'm not entirely sure if I did what you meant here, so it would be great if you could take a look at what I did

File: airflow_connector.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  else:
+    db = MySQLdb.connect(
+        host=CLOUDSQL_HOST,
+        user=CLOUDSQL_USER,
+        passwd=CLOUDSQL_PASSWORD,
+        db='airflow-db')
+
+  return db
+
+DB = connect_to_cloudsql()
```
Comment (hklai) [inline airflow_connector.py]: global variable is not a good idea

File: airflow_connector.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  Args:
+      request: the SQL request (string)
+
+  Returns:
+      The database result (tuple)
+  """
+  # db = connect_to_cloudsql()
+  cursor = DB.cursor()
+  cursor.execute(request)
+  response = cursor.fetchall()
```
Comment (hklai) [inline airflow_connector.py]: is this supposed to raise Error in case of failure?

File: rest_api.py (hunk 1 above)
Comment (hklai) [reply rest_api.py]: Can you do https://stackoverflow.com/questions/72852/how-to-do-relative-imports-in-python instead?

File: rest_api.py [hunk 9]
```diff
@@ -1,11 +1,13 @@
...
+import sys
+sys.path.append('../')
+from file_adapter import FileAdapter   # pylint: disable=g-import-not-at-top
+from filter_releases import filter_releases   # pylint: disable=g-import-not-at-top
+from filter_releases import sort   # pylint: disable=g-import-not-at-top
+from flask_restful import reqparse   # pylint: disable=g-import-not-at-top
+from flask_restful import Resource   # pylint: disable=g-import-not-at-top
+from airflow_connector import query_airflow  # pylint: disable=g-bad-import-order
 
 adapter = FileAdapter('fake_release_data.json', 'fake_task_data.json')
```
Comment (hklai) [inline rest_api.py]: A better abstraction is'

// 1. Builder paradigm
adapter = FileAdapter().loadRelease('fake_release_data.json').loadTasks('fake_task_data.json')

// 2. Old fashopm
adapter = FileAdapter()
adapter.loadRelease('fake_release_data.json')
adapter.loadTasks('fake_task_data.json')

However, why two files? Shouldn't the tasks own by the releases?

File: rest_api.py (hunk 4 above)
Comment (hklai) [reply rest_api.py]: This class should probably do connection and result set management.
i.e. this owns the connection(s) and close them when server exits, instead of using a global DB connection.

You may also want to verify that you can open multiple cursors on the same connection.. I remember some old jdbc drivers support only one at a time.

File: rest_api.py [hunk 10]
```diff
@@ -101,10 +103,22 @@ def get(self):
     return json.dumps(response)
 
 
+class AirflowDB(Resource):
+
+  def get(self):
+    parser = reqparse.RequestParser()
+    parser.add_argument('cm')
+    args = parser.parse_args()
+    data = query_airflow(str(args['cm']))
```
Comment (hklai) [inline rest_api.py]: how is the query constructed?

Comment (DomMarq) [reply rest_api.py]: What exactly do you mean? The query is in SQL query format

File: airflow_connector.py (hunk 6 above)
Comment (DomMarq) [reply airflow_connector.py]: The SQL database which we will be drawing all of the data from will always be "airflow-db" unless there is a change, most likely in the airflow source code

File: rest_api.py (hunk 4 above)
Comment (DomMarq) [reply rest_api.py]: If we are associating the DB connection with the class, then the server would have to open a new connection each time a request is made if we put it in a 'init' function which seems to me as if it negates the idea of having a consistent connection open.

Additionally, it seems that while my brief testing seems to allow for multiple cursors to be open on the same time, the documentation suggests closing the cursor after executing a query, which I have added

File: rest_api.py (hunk 9 above)
Comment (DomMarq) [reply rest_api.py]: With regards to having two files, I seem to remember that either you or Seb asked us to split them into two different sources, with the releases containing only references to the tasks which the release contained, in order to lessen the amount of data which we send to the release-ui homepage, as we don't need to have all of the task data present initially, especially considering the tasks is where a large portion of the data for a release resides

File: rest_api.py (hunk 4 above)
Comment (hklai) [reply rest_api.py]: DB connection should have proper scope. In your case you can open it when server starts and close it when server stops, but it (along with most things in any server) should not be global objects.

File: airflow_connector.py (hunk 6 above)
Comment (hklai) [reply airflow_connector.py]: better to move it to a constant and document it.

File: airflow_connector.py (hunk 8 above)
Comment (DomMarq) [reply airflow_connector.py]: Currently the frontend will display:
{"message": "Internal Server Error"}

and the server will raise this error: 
ProgrammingError: (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'showtables' at line 1")

However, I can implement a custom error if you wish

File: rest_api.py (hunk 9 above)
Comment (hklai) [reply rest_api.py]: They are different things actually....
1. Release dashboard shows a list of releases with some basic information like current state and last failed task, if any.
2. Release detail page shows all details about the release, including tasks.

Don't we already need info from #2 for #1 above? Separating them is mainly a UI concern, as we cannot dump too much info on screen at one time.

Assuming the release filter is sensible (i.e. last 7 days), sending all release + tasks from server to UI is not a lot of data. And if the result set is huge, pagination should kicks in.

File: rest_api.py (hunk 10 above)
Comment (hklai) [reply rest_api.py]: Who construct the SQL queries? airflow adapter?
And how does it derive filtering, grouping, and ordering conditions?

Comment (DomMarq) [reply rest_api.py]: Currently an airflow adapter has not been implemented, though eventually the airflow adapter will be responsible for constructing the airflow queries. The /airflowdb request handler which is currently implemented is there more for manual SQL queries rather than use by the server itself. (I implemented this initially because it can be very difficult to test whether things are working as expected in App Engine by itself, and this was a simple fix for me)

In the future I do not plan on using /airflowdb to handle requests from the frontend at all. Instead I plan to have /releases, /release, and /tasks all call the airflow adapter individually, and will pass in the requisite parameters to the adapter which will allow the adapter to construct the SQL query. These HTTP methods will then return the requested data which fits the parameters

File: rest_api.py (hunk 9 above)
Comment (DomMarq) [reply rest_api.py]: I believe that Emilia's PR changes the data structure significantly, making the concern of the particular data files somewhat irrelevant - however I believe that the same idea regarding holding only the task IDs in a release object as opposed to all of the task objects inside of release object holds true.

Additionally, the data which we will be receiving from the SQL database will be very similar to this to begin with, as the releases and task instances are in different tables. Obviously the adapter is there specifically to adapt the raw data from SQL to our own data structure.

Ultimately, it seems as if we have switched between how we store releases and tasks a few times, as you and @sebastienvas seem to have differing opinions on how release data should be represented. It would be great if we could agree on one before continuing.

File: airflow_connector.py (hunk 5 above)
Comment (DomMarq) [reply airflow_connector.py:1]: Where do I find it? Also a large portion of this was adapted from the [Cloud SQL / App Engine documentation](https://cloud.google.com/appengine/docs/standard/python/cloud-sql/using-cloud-sql-mysql), so I'm not too sure how that works either

File: airflow_connector.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+  CLOUDSQL_USER = 'root'
+  CLOUDSQL_PASSWORD = ''
+  CLOUDSQL_HOST = '35.193.234.53'
+  CLOUDSQL_DB = 'airflow-db'
+
+
+class AirflowDB(object):
+  """"Provides the methods which allow interaction with the Airflow SQL database."""  # pylint: disable=line-too-long
+
+  def __init__(self):
```
Comment (sebastienvas) [inline airflow_connector.py]: you should be passing everything as arguments in init not thru global variable. When you instantiate AirflowDB you can pass them via flags in your main.

File: app.yaml [hunk 12]
```diff
@@ -2,10 +2,20 @@ runtime: python27
 threadsafe: true
 api_version: 1
 
+env_variables:
```
Comment (sebastienvas) [inline app.yaml:5]: can you just pass flags to your script instead ?

File: main.py [hunk 13]
```diff
@@ -16,6 +16,7 @@
 API.add_resource(REST.release, '/release')
 API.add_resource(REST.labels, '/labels')
 API.add_resource(REST.tasks, '/tasks')
+API.add_resource(REST.airflowdb, '/airflowdb')
```
Comment (sebastienvas) [inline main.py]: Why do we need a API for this ?

Comment (DomMarq) [reply main.py]: This will eventually be removed, it's there right now for testing

File: rest_api.py [hunk 14]
```diff
@@ -101,10 +104,23 @@ def get(self):
     return json.dumps(response)
 
 
+class RestAirflowDB(Resource):
```
Comment (sebastienvas) [inline rest_api.py]: is this just for testing ?

Comment (DomMarq) [reply rest_api.py]: Yes

File: tests/rest_api_tests.py [hunk 15]
```diff
@@ -1,7 +1,9 @@
 """Unit Tests for Rest API."""
 
+import sys
 import unittest
-import rest_api
+sys.path.append('../')   # pylint diable=g-import-not-at-top
```
Comment (sebastienvas) [inline tests/rest_api_tests.py]: still don t understand why this is needed.

Comment (DomMarq) [reply tests/rest_api_tests.py]: Technically, it isn't. The only reason it is there is to allow the tests to be run from the tests/ folder, rather than having to move them out of the folder to the same directory as the file which is to be tested/imported. It is just there because I got tired of having to move the test file. If you would like, I can just remove it.

File: airflow_connector.py (hunk 5 above)
Comment (sebastienvas) [reply airflow_connector.py:1]: We don't have copyright yet. Let's add copyright at once in another PR maybe.

File: airflow_connector.py [hunk 16]
```diff
@@ -0,0 +1,55 @@
+"""Connects to the Cloud SQL database."""
+import os
+import MySQLdb
+
+
+class AirflowDB(object):
+  """"Provides the methods which allow interaction with the Airflow SQL database."""  # pylint: disable=line-too-long
+
+  def __init__(self, CLOUDSQL_CONNECTION_NAME, CLOUDSQL_USER, CLOUDSQL_PASSWORD, CLOUDSQL_HOST, CLOUDSQL_DB):
```
Comment (sebastienvas) [inline airflow_connector.py]: use lower case for parameters, and document the methods. What are the args and and what do they return.

File: airflow_connector.py [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+    """Connects to the Cloud SQL database."""
+    # When deployed to App Engine, the `SERVER_SOFTWARE` environment variable
+    # will be set to 'Google App Engine/version'.
+    if os.getenv('SERVER_SOFTWARE', '').startswith('Google App Engine/'):
```
Comment (sebastienvas) [inline airflow_connector.py]: don't look for env in the code, this gets ugly very fast. Pass in a boolean in your constructor instead.

Comment (sebastienvas) [reply airflow_connector.py]: actually even better pass in DB connect in your constructor, and you can construct it in you main().

File: main.py (hunk 13 above)
Comment (sebastienvas) [reply main.py]: Can you add a TODO then ?

Comment (DomMarq) [reply main.py]: Done

File: airflow_connector.py (hunk 17 above)
Comment (DomMarq) [reply airflow_connector.py]: I'm a little confused. Are you saying that I should construct the DB connection in main.py and then pass it in to the AirflowDB object when I initialize it?

**Comment by DomMarq** [comment]
Rebasing added a lot of changed files, but the only ones I have actually been working on and are submitting are airflow_connector, resources, and main

**Comment by sebastienvas** [comment]
Please fix conficts.

File: airflow_connector.py [hunk 18]
```diff
@@ -0,0 +1,67 @@
+"""Connects to the Cloud SQL database."""
+import os
+import MySQLdb
+
+
+class AirflowDB(object):
+  """"Provides the methods which allow interaction with the Airflow SQL database."""  # pylint: disable=line-too-long
+
+  def __init__(self, cloudsql_connection_name, cloudsql_user, cloudsql_password, cloudsql_host, cloudsql_db):
```
Comment (sebastienvas) [inline airflow_connector.py]: please pass in a db connect as I act in the previous review.

Comment (DomMarq) [reply airflow_connector.py]: I commented on the last time you said this, and I was a little confused as to where you want me to place the db connect constructor? And since you want me to essentially move the entire init function out of the object, should I just remove the object completely and just have a simpler query function, or leave the db object, and initialize it using the db connection?

Comment (sebastienvas) [reply airflow_connector.py]: Sorry I am just saying to pass the db connect in the constructor:
```python
def __init__(self, db_connection):
  self._db = db_connection.
```

And then you create the database connection in your main. That way you can get rid of the env variables.

File: airflow_connector.py [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+          db=cloudsql_db)
+
+    else:
+      db = MySQLdb.connect(
+          host=cloudsql_host,
+          user=cloudsql_user,
+          passwd=cloudsql_password,
+          db=cloudsql_db)
+
+    self.db = db
```
Comment (sebastienvas) [inline airflow_connector.py]: make this private.
```python
self._db
```

File: filter_releases.py [hunk 20]
```diff
@@ -0,0 +1,96 @@
+"""Helper Functions to be used for sorting and filtering."""
+from datetime import datetime
+
+
+NAME_DECR = 1
```
Comment (sebastienvas) [inline filter_releases.py]: you should probably have a boolean for ascending.

```python
class Sorting(object):
   BY_NAME_DECR = 1
   BY_CREATION_INCR = 2
   BY_LAST_MODIFIED = 3
   BY_LAST_ACTIVE = 4
```
and add the boolean:

```python
+def sort(releases, sort_method, reverse=True):

File: airflow_connector.py [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+    """Sends an SQL query to the airflow database.
+
+    Args:
+        request: the SQL request (string)
+
+    Returns:
+        The database result (tuple)
+    """
+    # db = connect_to_cloudsql()
+    print request
```
Comment (hklai) [inline airflow_connector.py]: remove these two lines?

File: app/scripts/controllers.js [hunk 22]
```diff
@@ -107,17 +107,34 @@ app.controller('MainController', ['$scope','$http','$location','$log', '$session
         offset = 0;
         $scope.$storage.currentPage = 1;
       }
+      // put the sorting into an enum and bool based format
+      var sortMethodDescending;
```
Comment (hklai) [inline app/scripts/controllers.js]: can you do just

sortMethodDescending = (($scope.$storage.sortMethod % 2) == 1)

?

What is $scope.$storage.sortMethod and what do its values entail?

File: filter_releases.py [hunk 23]
```diff
@@ -58,39 +48,39 @@ def filter_releases(releases, state, label, start_date, end_date, datetype):
...
 
-def sort(releases, sort_method):
+class Sorting(object):
+  BY_NAME = 1
+  BY_CREATION = 2
+  BY_LAST_MODIFIED = 3
+  BY_LAST_ACTIVE = 4
+
+
+def sort(releases, sort_method, descending):
```
Comment (hklai) [inline filter_releases.py]: Don't we just rely on the DB to sort and filter? What is this class for?

File: airflow_connector.py [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+  """"Provides the methods which allow interaction with the Airflow SQL database."""  # pylint: disable=line-too-long
+
+  def __init__(self, db_connection):
+    """Connects to the Cloud SQL database."""
+    # If I try to put this in main.py, then I get a circular import error when
+    # I try to import the connection/AiflowDB object into resources.py
+    # I could put it somewhere else, but for now it's here. It also seems that
+    # App Engine simply let me delete the env vars and everything is working finished
+
+    self._db = db_connection
```
Comment (hklai) [inline airflow_connector.py]: Should the connection be made inside instead?

File: resources.py [hunk 25]
```diff
@@ -1,12 +1,27 @@
...
+# from main import airflow_db
+# from main import db_connection
 
 adapter = FileAdapter('fake_data/fake_release_data.json', 'fake_data/fake_task_data.json')
 release_requests = {}
+# initiate a connection with the airflow database
+cloudsql_unix_socket = os.path.join('/cloudsql', 'istio-release-ui:us-central1:prod-airflow-snapshot-sandbox')
+db_connection = MySQLdb.connect(
+    unix_socket=cloudsql_unix_socket,
+    host='35.193.234.53',
```
Comment (hklai) [inline resources.py]: I suppose all these need to be parameterized?

File: resources.py [hunk 26]
```diff
@@ -19,9 +34,11 @@ def in_cache(args):
   state = str(int(args['state']))
   label = str(args['label'])
   sort_method = str(int(args['sort_method']))
+  descending = str(args['descending'])
 
   # create an autogenerated key for cache
-  key = start_date + end_date + datetype + state + label + sort_method
+  key = start_date + end_date + datetype + state + label + sort_method + descending
+  print key
```
Comment (hklai) [inline resources.py]: remove?

Comment (hklai) [reply resources.py]: all the other print statements too..

File: filter_releases.py [hunk 27]
```diff
@@ -58,39 +48,39 @@ def filter_releases(releases, state, label, start_date, end_date, datetype):
...
   """Sorts 'releases' according to 'sort_method'.
 
   Args:
     releases: array of release objects
-    sort_method: int representation of a sort sort_method
+    sort_method: int representation of a sort_method
+    descending: int representing bool descending
```
Comment (sebastienvas) [inline filter_releases.py]: please use reverse to make it consistent.

File: resources.py (hunk 25 above)
Comment (sebastienvas) [reply resources.py]: +1 That s why those should be in main. I am confused about the Resource class. Why do we have to use function pointers ? Must every Resource object have a constructor without parameters ?

File: resources.py [hunk 28]
```diff
@@ -112,10 +131,31 @@ def get(self):
...
+    parser.add_argument('end_date')
+    parser.add_argument('datetype')
+    parser.add_argument('state')
+    parser.add_argument('label')
+    parser.add_argument('sort_method')
+    parser.add_argument('limit')
+    parser.add_argument('offset')
+    args = parser.parse_args()
+    data = airflow_db.query(str(args['cm']))
+    print type(data)
```
Comment (sebastienvas) [inline resources.py]: use logging library instead of print statemens.

File: tests/rest_api_tests.py [hunk 29]
```diff
@@ -1,7 +1,9 @@
 """Unit Tests for Rest API."""
 
+import sys
 import unittest
-import rest_api
+sys.path.append('../')   # allows for importing of files not in the same directory | pylint: disable=line-too-long
```
Comment (sebastienvas) [inline tests/rest_api_tests.py:5]: do that in the __init__.py. @cabreraem has a fix in another PR.

File: app/scripts/controllers.js (hunk 22 above)
Comment (DomMarq) [reply app/scripts/controllers.js]: I tried to do something like that initially, but I was having some issues with sending a boolean over HTTP, and sending a bool in int format was a lot easier.

$scope.$storage.sortMethod is an int variable(I'm not sure why it's in that format, I think it has to do with the variable going between different controllers), and it contains the sort method enumerations

File: filter_releases.py (hunk 23 above)
Comment (DomMarq) [reply filter_releases.py]: This is mainly here for the file adapter, though it is possible that we will have to use this later to sort/filter the releases for certain cases. This is because, as I explained yesterday, we have to assemble some release data from the tasks, and so we most likely won't be able to push all the filtering to SQL.

The class is used for the sort method enumerations, per [this comment](https://github.com/istio-releases/release-ui/pull/31#pullrequestreview-136325971)

File: airflow_connector.py (hunk 24 above)
Comment (DomMarq) [reply airflow_connector.py]: What do you mean inside? Inside what?

File: filter_releases.py (hunk 23 above)
Comment (hklai) [reply filter_releases.py]: I think we probably should make those information available in xcom to facilitate sorting/filtering using sql directly. wdyt?

File: airflow_connector.py (hunk 24 above)
Comment (hklai) [reply airflow_connector.py]: Creating and initialing db connection inside this class, instead of taking a connection created outside.

This looks like the abstraction to handle all DB related stuff.

File: filter_releases.py (hunk 23 above)
Comment (DomMarq) [reply filter_releases.py]: That could certainly make things easier on the ui backend, at least for filtering. We could do something like this:
1. store the hard-to-get release info in xcom
2. query these vars using dag_id & execution_date
3. take the var values and use them to query certain releases

However, I don't think this approach will work for sorting, as there's currently nothing in the SQL database (that I'm aware of) which will allow for sorting of results from a table based on variables not contained in said table

File: airflow_connector.py (hunk 24 above)
Comment (DomMarq) [reply airflow_connector.py]: That was what I had initially, but Seb asked me to change it [here](https://github.com/istio-releases/release-ui/pull/31#discussion_r201759439)

File: tests/rest_api_tests.py (hunk 29 above)
Comment (DomMarq) [reply tests/rest_api_tests.py:5]: That's the plan, just waiting for it to be approved

File: resources.py (hunk 25 above)
Comment (DomMarq) [reply resources.py]: I'm not really sure what you mean by function pointers, so it would be great if you could explain that. With regards to the second question, as far as I can tell/have been able to figure out, yes it would seem so.

File: airflow_connector.py (hunk 24 above)
Comment (hklai) [reply airflow_connector.py]: @sebastienvas I suppose the airflow connector should abstract out all DB/airflow interactions, including dealing with the DB. What are the reasons why you want the connection object to be passed instead and have main create the connection?

I suppose this abstraction should have all the required info (i.e. user/pass/connection_info/etc) to be able to create connections. Also, connection object can turn bad due to network glitch (network resiliency is not always implemented in driver) and this abstraction may need to re-create db connection.

Not sure what "env vars" was referred to in your comment, but I agree that we should not rely on env vars here for sure. This can just take all required info as input params.

Comment (DomMarq) [reply airflow_connector.py]: >Also, connection object can turn bad due to network glitch (network resiliency is not always implemented in driver) and this abstraction may need to re-create db connection.
>
I've actually run into this problem a few times already during development/testing

File: resources.py (hunk 25 above)
Comment (DomMarq) [reply resources.py]: With regards to @hklai 's question, yes according to the docs I read I believe they do

File: main.py [hunk 30]
```diff
@@ -1,27 +1,47 @@
...
+
+my_resources = Releases(adapter)
 
 # adding resource endpoints to different urls
-API.add_resource(REST.releases, '/releases')
-API.add_resource(REST.release, '/release')
-API.add_resource(REST.branches, '/branches')
-API.add_resource(REST.types, '/types')
-API.add_resource(REST.tasks, '/tasks')
+API.add_resource(resources.releases, '/releases', resource_class_kwargs={'adapter': adapter})
```
Comment (sebastienvas) [inline main.py]: I think you meants resource.Releases ad same thing for the other one.

File: main.py [hunk 31]
```diff
@@ -1,27 +1,47 @@
...
 
 # creating the Flask application
 APP = Flask(__name__)
 API = Api(APP)
+resources = Resources()
+
 
-REST = Resources()
+adapter = FileAdapter('fake_data/fake_release_data.json', 'fake_data/fake_task_data.json')
+cloudsql_unix_socket = os.path.join('/cloudsql', 'istio-release-ui:us-central1:prod-airflow-snapshot-sandbox')
```
Comment (sebastienvas) [inline main.py]: Use flags https://docs.python.org/3.3/library/argparse.html

Comment (sebastienvas) [reply main.py]: I still feel that creating the db connection here, and passing it down allows us to test the core more easily.

File: main.py (hunk 30 above)
Comment (DomMarq) [reply main.py]: It actually doesn't work if it is capitalized, since it refers to the Resources class in resources.py. If you want, I can change it to capitals though.

File: resources.py [hunk 32]
```diff
@@ -1,12 +1,21 @@
...
+from flask_restful import Resource
 from file_adapter import FileAdapter
-from filter import filter_releases, sort
-from flask_restful import Resource, reqparse
+
+
+# from main import airflow_db
+# from main import db_connection
 
 adapter = FileAdapter('fake_data/fake_release_data.json', 'fake_data/fake_task_data.json')
```
Comment (sebastienvas) [inline resources.py]: you should be able to remove this

File: main.py (hunk 31 above)
Comment (DomMarq) [reply main.py]: It might be better for testing(I can't really say), but creating it in the adapter allows for me to easily reconnect if the connection is lost, without raising an error and cancelling a query

File: resources.py (hunk 32 above)
Comment (DomMarq) [reply resources.py]: Done

File: resources.py [hunk 33]
```diff
@@ -20,9 +29,10 @@ def in_cache(args):
   branch = str(args['branch'])
   type = str(args['type'])
   sort_method = str(int(args['sort_method']))
+  descending = str(args['descending'])
```
Comment (sebastienvas) [inline resources.py]: reverse ?

File: resources.py [hunk 34]
```diff
@@ -20,9 +29,10 @@ def in_cache(args):
   branch = str(args['branch'])
   type = str(args['type'])
```
Comment (sebastienvas) [inline resources.py]: type is keyword in python. Please call it release_type or something like this

File: main.py (hunk 31 above)
Comment (DomMarq) [reply main.py]: Additionally, I don't believe that App Engine will be able to make use of these flags, so I assume that you want me to make the current values default?

**Review by sebastienvas — APPROVED** [review]
